contracts/access/AccessControlAdminRules.sol

```diff
@@ -0,0 +1,220 @@
+     * functions.
+     *
+     * For other roles, see {AccessControl-renounceRole}.
+     *
+     * NOTE: Renouncing `DEFAULT_ADMIN_ROLE` will leave the contract without an owner,
+     * thereby disabling any functionality that is only available to the default admin, and the
+     * possibility of reassigning a non-administrated role.
+     */
+    function renounceRole(bytes32 role, address account) public override(IAccessControl, AccessControl) {
+        if (role == DEFAULT_ADMIN_ROLE) {
+            require(
+                pendingAdmin() == address(0) && _adminTransferIsUnlocked(),
+                "AccessControl: admin can only renounce in two delayed steps"
+            );
+        }
+        super.renounceRole(role, account);
+    }
+
+    /**
+     * @dev See {AccessControl-grantRole}. Reverts for `DEFAULT_ADMIN_ROLE`.
+     */
+    function grantRole(
+        bytes32 role,
+        address account
+    ) public override(IAccessControl, AccessControl) onlyRole(getRoleAdmin(role)) {
+        require(role != DEFAULT_ADMIN_ROLE, "AccessControl: can't directly grant admin role");
+        super.grantRole(role, account);
+    }
+
+    /**
+     * @dev See {AccessControl-revokeRole}. Reverts for `DEFAULT_ADMIN_ROLE`.
+     */
+    function revokeRole(
+        bytes32 role,
+        address account
+    ) public override(IAccessControl, AccessControl) onlyRole(getRoleAdmin(role)) {
+        require(role != DEFAULT_ADMIN_ROLE, "AccessControl: can't directly revoke admin role");
+        super.revokeRole(role, account);
+    }
+
+    /**
+     * @dev See {AccessControl-_setRoleAdmin}. Reverts for `DEFAULT_ADMIN_ROLE`.
+     */
+    function _setRoleAdmin(bytes32 role, bytes32 adminRole) internal override {
+        require(role != DEFAULT_ADMIN_ROLE, "AccessControl: can't override admin's admin");
+        super._setRoleAdmin(role, adminRole);
+    }
+
+    /**
+     * @dev Grants `role` to `account`.
+     *
+     * For `DEFAULT_ADMIN_ROLE`, it only allows granting if there isn't already a role's owner
+     * or if the role has been previously renounced.
+     *
+     * For other roles, see {AccessControl-renounceRole}.
+     *
+     * NOTE: Exposing this function through another mechanism may make the
+     * `DEFAULT_ADMIN_ROLE` assignable again. Make sure to guarantee this is
+     * the expected behavior in your implementation.
+     */
+    function _grantRole(bytes32 role, address account) internal override {
+        if (role == DEFAULT_ADMIN_ROLE) {
+            require(owner() == address(0), "AccessControl: admin already granted");
+            _currentAdmin = account;
+        }
+        super._grantRole(role, account);
+    }
+
+    /**
+     * @dev See {AccessControl-_revokeRole}.
+     */
+    function _revokeRole(bytes32 role, address account) internal override {
+        if (role == DEFAULT_ADMIN_ROLE) {
+            delete _currentAdmin;
+        }
+        super._revokeRole(role, account);
+    }
```

Comment:
when doing 

```solidity
        _revokeRole(DEFAULT_ADMIN_ROLE, owner());
        _grantRole(DEFAULT_ADMIN_ROLE, pendingAdminOwner);
```

The current `_currentAdmin` is first set to 0, then written to. That is gas consuming (maybe its ok).
Also, if someone was to change that to

```solidity
        _grantRole(DEFAULT_ADMIN_ROLE, pendingAdminOwner);
        _revokeRole(DEFAULT_ADMIN_ROLE, owner());
```

things would break in a non obvious way. I don't have a better solution, but that is dangerous.

Reply:
This function is also not `virtual`, so doing `_currentAdmin` instead of owner can be the solution to writing `_grantRole` first.

The only other way it can break is by overriding `_grantRole`, but that itself is also dangerous so I wouldn't expect people doing it.

> That is gas consuming (maybe its ok).

Here are some benchmarks:

```bash
## `_revokeRole` -> `_grantRole`
Running 1 test for test/access/AccessControlAdminRules.t.sol:AdminRules
[PASS] testAccept(address) (runs: 10000, μ: 53573, ~: 53558)

## `_grantRole` -> `_revokeRole`
Running 1 test for test/access/AccessControlAdminRules.t.sol:AdminRules
[PASS] testAccept(address) (runs: 10000, μ: 51334, ~: 51318)
```

So, there IS a difference, but I needed to remove the `require(owner() == address(0), "AccessControl: admin already granted");` check in both to make the second one pass.

Reply:
I just noticed that `_grantRole` requires that `owner()` is `address(0)` ... so that is good.

Reply:
| op | gas | refund |
|-|-|-|
SSTORE currentAdmin 0 | 2100 + 2900 | +4800 
SSTORE role[currentAdmin] 0 | 2100 + 2900 | +4800
SSTORE currentAdmin newAdmin | 100 | - 4800
SSTORE role[newAdmin] 1 | 2100 + 20000 | 

The total gas is 32200, the total refund is 4800. The refund is less than half of the total gas so it gets applied, resulting in 27400 final gas.

If we simply did `SSTORE currentAdmin newAdmin` in 1 step, the cost of that operation would be 5000 gas. In 2 steps, the cost is 5100 gas, so we would only be saving 100 gas.

I think the overhead of the additional step is fine.